File: src/main/java/com/deneb/client/command/Command.java
package com.deneb.client.command;

import com.deneb.client.utils.Wrapper;
import net.minecraft.client.Minecraft;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by B_312 on 01/15/21
 */
public abstract class Command {
    private final String command;
    private final String description;

    public static final Minecraft mc = Wrapper.mc;

    public Command() {
        this.command = getAnnotation().command();
        this.description = getAnnotation().description();
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface Info {
        String command();

        String description() default "";
    }

    private Info getAnnotation() {
        if (getClass().isAnnotationPresent(Info.class)) {
            return getClass().getAnnotation(Info.class);
        }
        throw new IllegalStateException("No Annotation on class " + this.getClass().getCanonicalName() + "!");
    }

    public abstract void onCall(String s, String... args);

    public abstract String getSyntax();

    public String getDescription() {
        return description;
    }

    public String getCommand() {
        return command;
    }
}

File: src/main/java/club/deneb/client/command/Command.kt
package club.deneb.client.command

import net.minecraft.client.Minecraft

/**
 * Created by B_312 on 01/15/21
 */
open class Command {
    val command: String
    val description: String

    @kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
    annotation class Info(val command: String, val description: String = "")

    private val annotation: Info
        get() {
            if (javaClass.isAnnotationPresent(Info::class.java)) {
                return javaClass.getAnnotation(Info::class.java)
            }
            throw IllegalStateException("No Annotation on class " + this.javaClass.canonicalName + "!")
        }

    open fun onCall(s: String, vararg args: String){}

    companion object {
        val mc: Minecraft = Minecraft.getMinecraft()
    }

    init {
        command = annotation.command
        description = annotation.description
    }

    open fun getSyntax(): String {
        return ""
    }
}